File: src-tauri/src/commands/editor_sync.rs
//! 本地编辑器同步命令：探测编辑器、打开/列出/关闭同步会话。
//!
//! 前端只传结构化标识（session_id、远程路径、editor_id）——路径只进 SFTP
//! 通道，绝不拼 shell 命令，与 `sftp.rs` 同一安全定位。

use tauri::{AppHandle, Emitter, State};

use super::record_audit;
use crate::editor_sync::{
    close_sync_session, find_editor, list_editors, open_sync_session, EditorSyncEventPayload,
    SyncSessionInfo, EDITOR_SYNC_EVENT,
};
use crate::state::AppState;

/// 探测本机可用的编辑器（VS Code / Cursor / Windsurf / Trae / CodeBuddy）。
/// 未安装的返回 `available: false`，前端不展示。
#[tauri::command]
pub async fn editor_list_available() -> Vec<crate::editor_sync::EditorInfo> {
    // 探测是纯文件存在性检查（微秒级），直接同步算完。
    list_editors()
}

/// 打开同步会话：下载远程文件/目录到本地工作区并用指定编辑器打开。
/// 之后编辑器每次保存都会自动回传服务器。
#[tauri::command]
pub async fn editor_sync_open(
    app: AppHandle,
    state: State<'_, AppState>,
    session_id: String,
    remote_path: String,
    editor_id: String,
) -> Result<SyncSessionInfo, String> {
    // 网络 I/O 之前完成所有校验。
    crate::editor_sync::validate_remote_path(&remote_path).map_err(|error| error.to_string())?;
    let editor = find_editor(&editor_id)
        .map(|(name, path)| (editor_id.clone(), name, path))
        .ok_or_else(|| format!("未检测到编辑器 {editor_id}，无法打开"))?;

    let info = open_sync_session(&app, &state.editor_syncs, &state.ssh, &session_id, &remote_path, editor)
        .await
        .map_err(|error| error.to_string())?;
    record_audit(&state, "editor_sync_open", None, None, &format!("{} → {}", info.remote_path, info.editor_name));
    Ok(info)
}

/// 关闭同步会话：停止监听与上传、清理本地临时工作区。
#[tauri::command]
pub async fn editor_sync_close(
    app: AppHandle,
    state: State<'_, AppState>,
    sync_id: String,
) -> Result<SyncSessionInfo, String> {
    let closed = close_sync_session(&state.editor_syncs, &sync_id)
        .await
        .map_err(|error| error.to_string())?;
    record_audit(&state, "editor_sync_close", None, None, &closed.remote_path);
    let _ = app.emit(EDITOR_SYNC_EVENT, EditorSyncEventPayload::upsert(closed.clone()));
    Ok(closed)
}

/// 列出同步会话（`session_id` 提供时只返回该 SSH 会话的）。
/// 面板打开/重挂载时调用，恢复状态条的显示。
#[tauri::command]
pub async fn editor_sync_list(
    state: State<'_, AppState>,
    session_id: Option<String>,
) -> Result<Vec<SyncSessionInfo>, String> {
    Ok(match session_id {
        Some(id) if !id.is_empty() => state.editor_syncs.snapshot_for_session(&id).await,
        _ => state.editor_syncs.snapshot().await,
    })
}

File: src/api/types/editor-sync.ts
/**
 * 本地编辑器同步域类型 —— 远程文件副本 → 本地编辑器（VSCode 等）打开 →
 * 保存自动 SFTP 回传。命令：`editor_sync_open/close/list`、
 * `editor_list_available`；事件：`editor-sync-update`
 * （src-tauri/src/editor_sync/，字段 camelCase、枚举值 snake_case 逐字一致）。
 */

/** 一个本机探测到的编辑器（未安装时 `available: false`）。 */
export interface EditorInfo {
  id: string;
  name: string;
  available: boolean;
  path: string | null;
}

/** 同步范围：单文件或整个目录。 */
export type EditorSyncScope = "file" | "directory";

/**
 * 会话状态：`starting` 极短窗口；`error` = 最近一次保存未同步（下一次保存
 * 成功自动回 active）；`closed` = 用户主动关闭后的终态。
 */
export type EditorSyncStatus = "starting" | "active" | "error" | "closed";

/** 一条同步会话的完整状态（事件载荷与命令返回共用）。 */
export interface SyncSessionInfo {
  id: string;
  /** 所属 SSH 会话。 */
  sessionId: string;
  scope: EditorSyncScope;
  /** 服务器上的根（文件或目录的绝对路径）。 */
  remotePath: string;
  /** 本地工作区绝对路径（单文件模式下是文件自身）。 */
  localPath: string;
  editorId: string;
  editorName: string;
  status: EditorSyncStatus;
  /** `error` 状态下的失败原因；其余状态为 null。 */
  message: string | null;
  /** 累计成功同步次数。 */
  syncCount: number;
  /** 最近一次成功同步的毫秒时间戳；0 = 从未。 */
  lastSyncAt: number;
  openedAt: number;
}

/** `editor-sync-update` 事件载荷。 */
export interface EditorSyncEventPayload {
  kind: string;
  session: SyncSessionInfo;
}
